strapi/strapi
# [Custom fields] Add default value to options

### What does it do?
- Registers every valid type of custom field for testing in `examples/getstarted` (this will break the admin on 4000 if you add a custom field and then try to access it in the content manager, but I think it is nice to have them to make sure they work in the CTB)
- Handles a `defaultValue` key on custom field option objects so that a defaultValue can be set for the option
- Validates the `name` on custom field options object (allowed root level type or prefixed with 'options')

### Why is it needed?

- Custom field developers should be able to provide default values for their options

### How to test it?

- Go to any content-type and add new field, select custom
- Choose color picker
- The color picker now adds a default value for `regex`, check it is pre-filled in the form modal
- Save your new field, check the schema json and confirm it was saved as follows, regex should be on the root of the attribute object
```
"test": {
  "pluginOptions": {
    "i18n": {
      "localized": true
    }
  },
  "type": "customField",
  "regex": "^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$",
  "customField": "plugin::color-picker.color"
}
```
- Go to `packages/plugins/color-picker/admin/src/index.js`, and add the following object to the `advanced` array of the custom field registration
```
  {
    intlLabel: {
      id: getTrad('test'),
      defaultMessage: 'Format',
    },
    name: 'options.format',
    type: 'text',
    defaultValue: 'hex',
    description: {
      id: getTrad('test'),
      defaultMessage: 'This is just a test',
    },
  }
```
- Do the same test as you did for `regex`, format should be on an options object
```
"test": {
  "pluginOptions": {
    "i18n": {
      "localized": true
    }
  },
  "type": "customField",
  "regex": "^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$",
  "options": {
    "format": "hex"
  },
  "customField": "plugin::color-picker.color"
}
```
- Go back to `packages/plugins/color-picker/admin/src/index.js` and change `name: 'options.format'` to  `name: 'format'`, save and go back to the admin, you should have an error.
